haskell/test/Model/BaseDAO/Types.hs

```diff
@@ -0,0 +1,236 @@
+-- SPDX-FileCopyrightText: 2021 TQ Tezos
+-- SPDX-License-Identifier: LicenseRef-MIT-TQ
+
+{-# OPTIONS_GHC -Wno-orphans #-}
+{-# OPTIONS_GHC -Wno-missing-signatures #-}
+
+module Model.BaseDAO.Types
+  ( ModelT (..)
+  , ModelState (..)
+  , SimpleContract (..)
+  , SimpleOperation (..)
+  , runModelT
+  , getStore
+  , getConfig
+  , modifyStore
+  , addOperations
+  , getEntrypoint
+  , mkSimpleOperation
+  , (-$$?)
+
+  , ModelCall (..)
+  , ModelError (..)
+  , contractErrorToModelError
+  )where
+
+import Universum
+
+import Control.Monad.Except (MonadError, throwError)
+import Data.Typeable (cast)
+import qualified Data.Map as M
+import Fmt
+
+import Michelson.Typed.Value
+import Michelson.Interpret
+import Michelson.Text
+import qualified Michelson.Typed as T
+import Lorentz hiding (not, cast, get)
+import Tezos.Address (ContractHash(..))
+import Tezos.Core
+import qualified Lorentz.Contracts.Spec.FA2Interface as FA2
+
+import Ligo.BaseDAO.Types
+
+-- | Transformer used in haskell implementation of BaseDAO
+newtype ModelT a = ModelT
+  { unModelT :: (ExceptT ModelError $ State ModelState) a
+  } deriving newtype (Functor, Applicative, Monad, MonadState ModelState, MonadError ModelError)
+
+-- | A type that keep track of blockchain state
+data ModelState = ModelState
+  { msFullStorage :: FullStorage
+  , msOperations :: [SimpleOperation]
```

Comment:
I'm not sure that I understand what's the point of this.

I mean, if we cannot compare the operations with the interpreter and we need to check the storage of auxiliary contract, why do we store simple operations? Can't we directly execute them in `ModelT` and update the `msContracts`?

Reply:
The problem with executing them in `ModelT` is that, we have to reset the auxiliary contract storage, if an error occurred. That would be pretty tricky to do. 

As a result, operations are only executed at the end of a finished call via as part of `runModelT`. Let me know if what you think about this.

Reply:
AFAIK we have to reset everything either way, in the case a transfer from the DAO to another contract fails during the execution of an entrypoint the whole call fails and everything should be reset. Seems related to [this](https://github.com/tqtezos/baseDAO/pull/282#discussion_r662457157)